lib/ecto/repo/queryable.ex

```diff
@@ -485,4 +536,28 @@ defmodule Ecto.Repo.Queryable do
       query: query,
       message: "expected a from expression with a schema"
   end
+
+  defp assert_structs!(structs) when is_list(structs) do
+    case Enum.reject(structs, &is_struct?/1) do
+      [] ->
+        :ok
+
+      rejected ->
+        raise ArgumentError, "Expected a struct or a list of structs, received #{inspect(rejected)}"
+    end
+  end
+
+  defp is_struct?(%{__struct__: _, __meta__: _}), do: true
+  defp is_struct?(_), do: false
```

Comment:
We typically use `is_` or `?` but not both.  In this case, since it is not a guard, `?` works better. :)

```suggestion
  defp schema?(%{__struct__: _, __meta__: _}), do: true
  defp schema?(_), do: false
```

Reply:
@josevalim I have a conceptual question: does `Map.put(%{}, :__struct__, ...)` imply that the result is a struct?

In my conception, matching on `%_{__meta__: _}` would be a little bit more semantic in this regard, but I really don't know if there is a practical difference

Reply:
This is a subset match, so you can match as above, but you shouldn’t build the struct manually as you may miss some fields. In any case your syntax is definitely nicer!